internal/controlplane/handlers_repositories.go

```diff
@@ -491,28 +455,62 @@ func (s *Server) ListRemoteRepositoriesFromProvider(
 	return out, nil
 }
 
-func (s *Server) deleteRepositoryAndWebhook(
+// TODO: this probably can probably be used elsewhere
+// returns project ID and github client - this flow of code exists in multiple
+// places
+func (s *Server) getProjectIDAndClient(
 	ctx context.Context,
-	repo db.Repository,
-	provider db.Provider,
-) error {
-	tx, err := s.store.BeginTransaction()
+	request HasProtoContext,
+) (uuid.UUID, v1.GitHub, error) {
+	projectID := getProjectID(ctx)
+
+	provider, err := getProviderFromRequestOrDefault(ctx, s.store, request, projectID)
 	if err != nil {
-		return status.Errorf(codes.Internal, "error deleting repository")
+		return uuid.Nil, nil, providerError(err)
 	}
-	defer s.store.Rollback(tx)
 
-	qtx := s.store.GetQuerierWithTransaction(tx)
-	if err := qtx.DeleteRepository(ctx, repo.ID); err != nil {
-		return status.Errorf(codes.Internal, "error deleting repository: %v", err)
+	pbOpts := []providers.ProviderBuilderOption{
+		providers.WithProviderMetrics(s.provMt),
+		providers.WithRestClientCache(s.restClientCache),
 	}
 
-	if err := s.deleteWebhookFromRepository(ctx, provider, repo); err != nil {
+	p, err := providers.GetProviderBuilder(ctx, provider, s.store, s.cryptoEngine, pbOpts...)
+	if err != nil {
+		return uuid.Nil, nil, status.Errorf(codes.Internal, "cannot get provider builder: %v", err)
+	}
+
+	if !p.Implements(db.ProviderTypeGithub) {
+		return uuid.Nil, nil, util.UserVisibleError(
+			codes.InvalidArgument,
+			"provider %s is not supported for github webhook",
+			p.GetName(),
+		)
+	}
+
+	client, err := p.GetGitHub()
+	if err != nil {
+		return uuid.Nil, nil, status.Errorf(codes.Internal, "error creating github provider: %v", err)
+	}
+
+	return projectID, client, nil
+}
+
+// covers the common logic for the two varieties of repo deletion
+func (s *Server) deleteRepository(
+	ctx context.Context,
+	request HasProtoContext,
+	deletionMethod func(v1.GitHub, uuid.UUID) error,
+) error {
+	projectID, client, err := s.getProjectIDAndClient(ctx, request)
+	if err != nil {
 		return err
 	}
 
-	if err := s.store.Commit(tx); err != nil {
```

Comment:
The old code deleted the repo from the DB, then deleted the webhook. The new code deletes the webhook, and then deletes from the DB. Since the deletion of the webhook is idempotent, I decided not to use a transaction.

Reply:
+1 to deleting the webhook first.  Otherwise, we could delete our record that the webhook existed before we'd succeeded at deleting the webhook itself.